torch/optim/_multi_tensor/adagrad.py

```diff
@@ -0,0 +1,114 @@
+            raise ValueError("Invalid lr_decay value: {}".format(lr_decay))
+        if not 0.0 <= weight_decay:
+            raise ValueError("Invalid weight_decay value: {}".format(weight_decay))
+        if not 0.0 <= initial_accumulator_value:
+            raise ValueError("Invalid initial_accumulator_value value: {}".format(initial_accumulator_value))
+        if not 0.0 <= eps:
+            raise ValueError("Invalid epsilon value: {}".format(eps))
+
+        defaults = dict(lr=lr, lr_decay=lr_decay, eps=eps, weight_decay=weight_decay,
+                        initial_accumulator_value=initial_accumulator_value)
+        super(Adagrad, self).__init__(params, defaults)
+
+        for group in self.param_groups:
+            for p in group['params']:
+                state = self.state[p]
+                state['step'] = 0
+                state['sum'] = torch.full_like(p, initial_accumulator_value, memory_format=torch.preserve_format)
+
+    def share_memory(self):
+        for group in self.param_groups:
+            for p in group['params']:
+                state = self.state[p]
+                state['sum'].share_memory_()
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        """Performs a single optimization step.
+
+        Args:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        for group in self.param_groups:
+            params_with_grad = []
+            grads = []
+            state_sums = []
+            state_steps = []
+
+            has_sparse_grad = False
+            for p in group['params']:
+                if p.grad is not None:
+                    if p.grad.is_sparse:
+                        has_sparse_grad = True
+                    params_with_grad.append(p)
+                    grads.append(p.grad)
+                    state = self.state[p]
+                    state_sums.append(state['sum'])
+                    # update the steps for each param group update
+                    state['step'] += 1
+                    # record the step after step update
+                    state_steps.append(state['step'])
+
+            weight_decay = group['weight_decay']
+            lr = group['lr']
+            lr_decay = group['lr_decay']
+
+            if weight_decay != 0:
+                if has_sparse_grad:
+                    raise RuntimeError("weight_decay option is not compatible with sparse gradients")
+                torch._foreach_add_(grads, params_with_grad, alpha=weight_decay)
+
+            minus_clr = [-lr / (1 + (step - 1) * lr_decay) for step in state_steps]
+
+            if has_sparse_grad:
+                # sparse is not supported by multi_tensor. Fall back to optim.adagrad
+                # implementation.
+                F.adagrad(params_with_grad,
```

Comment:
We don't call single tensor functional optimizer from multi tensor API yet. Do you think you can do it without calling this function ? I can suggest you two ways:

1. We do have multi tensor functional optimizers:  https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/_functional.py , you can add multi-tensor implementation here and call this new function.

2. You can avoid add your code here yourself, without calling this function.